lib/puppet/util/windows/service.rb

```diff
@@ -273,18 +283,25 @@ def start(service_name)
     # @param [:string] service_name name of the service to stop
     def stop(service_name)
       open_service(service_name, SC_MANAGER_CONNECT, SERVICE_STOP | SERVICE_QUERY_STATUS) do |service|
+        # A truthy value return from pending transision indicates the service is now in
+        # SERVICE_STOPPED, a falsey value indicates the service is in another state
+        if wait_for_pending_transition(service, [SERVICE_STOP_PENDING], SERVICE_STOPPED)
+          # just return if the service is already stopped, otherwise continue
+          # All other states other than STOP_PENDING and STOPPED are required to accept the
+          # stop control, so executtion should succeed
+          return
+        end
         FFI::MemoryPointer.new(SERVICE_STATUS.size) do |status_ptr|
           status = SERVICE_STATUS.new(status_ptr)
-          # don't attempt to fail here if the service isn't started because windows error codes
-          # are likely more informative than ours and a failed call to ControlService will produce
-          # those errors
-          wait_for_pending_transition(service, SERVICE_START_PENDING, SERVICE_RUNNING)
           if ControlService(service, SERVICE_CONTROL_STOP, status) == FFI::WIN32_FALSE
             raise Puppet::Util::Windows::Error.new(_("Failed to send stop control to service, current state is %{current_state}. Failed with") % { current_state: status[:dwCurrentState].to_s })
           end
-          unless wait_for_pending_transition(service, SERVICE_STOP_PENDING, SERVICE_STOPPED)
-            raise Puppet::Error.new(_("Failed to stop the service, after calling ControlService the service is not in SERVICE_STOP_PENDING or SERVICE_STOPPED"))
-          end
+        end
+        # We've seen instances *cough* nssm *cough* where the service remains in RUNNING before the service
```

Comment:
I don't think NSSM is strictly doing the "wrong" thing, because `RUNNING => STOPPED` is a valid state transition based on the transition diagram here: https://docs.microsoft.com/en-us/windows/desktop/services/service-status-transitions (and it could be that `RUNNING => STOPPED` is a "fast" transition for `pxp-agent` so that `STOP_PENDING` is unnecessary, but just not fast enough for Puppet to catch it OR it could take some time for the service to transition to `STOP_PENDING` since `ControlService` is asynchronous OR some other stuff). I wonder if it would be better to refactor `wait_for_pending_transition` into two methods: `wait_for_possible_state_transition`, `wait_for_state_transition`, where their interface would be something like:
```
wait_for_possible_state_transition(service, valid_initial_states, final_states)
wait_for_state_transition(service, valid_initial_states, final_states)
```
The difference between these two methods is that `wait_for_possible_state_transition` would pass-thru if the service’s initial state does not match any of the valid_initial_states. Here’s how they would work in practice, using the `Stop` method as an example:

```
// This would ensure that if we’re in SERVICE_START_PENDING, we wait for it to get to SERVICE_RUNNING and then
// fail if it doesn’t (which is what we want). If we’re in any other initial state like SERVICE_PAUSED, SERVICE_CONTINUE_PENDING,
// we pass-thru which is fine b/c we’re safe to stop the service in those states (since the service thread’s already been
// created).
wait_for_possible_state_transition(service, [SERVICE_START_PENDING], SERVICE_RUNNING)

// Here, a service cannot be in the SERVICE_START_PENDING state.

// Send the stop signal
ControlService

// The valid_initial_states here should be all of the states that we can get to SERVICE_STOP_PENDING or SERVICE_STOPPED
// from, based on the state transition diagram. These are all of the states except for SERVICE_STOP_PENDING and SERVICE_STOPPED.
// This routine is here in case e.g. it takes some time to go from SERVICE_RUNNING => SERVICE_STOP_PENDING or if a service
// goes to SERVICE_STOPPED directly.
wait_for_possible_state_transition(service, [SERVICE_RUNNING, SERVICE_PAUSED, ...], [SERVICE_STOP_PENDING, SERVICE_STOPPED])

// Here, a service is either in the SERVICE_STOP_PENDING or SERVICE_STOPPED states.

// Wait for SERVICE_STOP_PENDING => SERVICE_STOPPED
wait_for_state_transition(service, [SERVICE_STOP_PENDING], SERVICE_STOPPED)
```

The logic for wait_for_possible_state_transition and wait_for_state_transition would be similar. Something like:
```
wait_for_state_transition(service, valid_initial_states, final_states)
  service_state = get_service_state(service)
  if final_states.include?(service_state)
    return
  end
  unless valid_initial_states.include?(service_state)
    // in wait_for_possible_state_transition, we would return here instead
    raise exception
  end
  initial_state = service_state
  loop do
    // Do the transition logic here. Note that if initial_state is a pending state, we would use the waitHint (if provided).
    // Otherwise, I think it is safe to use a default timeout here for e.g. SERVICE_RUNNING, SERVICE_PAUSED, etc.
    // We would fail if the service is not in one of the final states here.
  end
```

Reply:
Yeah, I didn't think about it but you are right, RUNNING -> STOPPED is valid.

@ekinanp I don't see a reason to split the method in to two separate methods. From what you described I'm already 'passing through' the wait function by just ignoring the return value. The wait function _only_ fails in the scenario that the service is in one of the listed pending states and the wait hint has elapsed. In any other case (i.e. if the service is in a state other than one you provided to wait for) it would return a boolean: true if the service is in the final state and false in any other case.

Thus in stop you just ignore the return value from the wait function and wait on SERVICE_START_PENDING, if the service is in SERVICE_START_PENDING it will wait, otherwise you will pass through because stop just ignores the return value of that wait call.

Reply:
I do likely need to update the last wait call in the stop function to basically take any state (except start pending) though.

Reply:
Oh, and stop is missing a wait transition on SERVICE_START_PENDING, whoops

Reply:
@McdonaldSeanp I keep forgetting the Boolean return, so that makes sense. I think having the `wait_for_possible_state_transition` is useful to make it easier to reason about what's going on here. The name implies that the transition we're waiting for is not something that's always going to happen, but a possibility we need to account for. `wait_for_state_transition` is a transition we have to wait for no matter what. Using the same method to handle a `possible` vs. `required` transition requires an extra amount of (unnecessary) thinking to differentiate between the two cases. Both methods should, however, preserve the Boolean return, just `wait_for_state_transition` throws an exception for an invalid initial state. Then we can reason about the code as e.g.
```
if wait_for_possible_state_transition(service, SERVICE_STOP_PENDING, SERVICE_STOPPED)
  return
end

wait_for_possible_state_transition(service, SERVICE_START_PENDING, SERVICE_RUNNING)
etc.
```

I still think you should have two waits at the end:
```
wait_for_possible_state_transition(service, [SERVICE_RUNNING, SERVICE_PAUSED, ...], [SERVICE_STOP_PENDING, SERVICE_STOPPED])
wait_for_state_transition(service, [SERVICE_STOP_PENDING], SERVICE_STOPPED)
```

The reason is b/c there could still be some delay for the service to go to `STOP_PENDING` even after the control signal. The single wait assumes that the transition is instantaneous.

Reply:
@ekinanp I pushed an update to add acceptable wait states to the final wait on the stop method. does the stop function make any more sense?

Reply:
@McdonaldSeanp That's pretty hacky. `RUNNING`, `PAUSED` are not pending states and should not be treated as such. Also, another issue is the `wait_hint`. `wait_hint` only makes sense in that final step when it is from `STOP_PENDING => STOPPED`. You don't want to use the same `wait_hint` that `CONTINUE_PENDING` provides, for example, when transitioning from `CONTINUE_PENDING => STOPPED` or `CONTINUE_PENDING => STOP_PENDING` because that `wait_hint` is only valid for `CONTINUE_PENDING => RUNNING`. That's another big reason why the two separate waits is better, because then before the last wait, your post-condition is that the service state is either `STOP_PENDING` or `STOPPED`, whereby the last wait will work correctly so that the wait_hint provided by `STOP_PENDING` will make sense at that last wait.

In essence, I think restructuring `wait_for_pending_transition` as `wait_for_state_transition` makes things simpler, more explicit, and consistent with the Windows service state transition diagram.

I don't have strong feelings about the timeout, so your reasoning makes sense there.

My point about the second method makes more sense if `wait_for_pending_transition` is restructured as `wait_for_state_transition`, so let's talk about that more once we agree upon this.

Reply:
Also, you cannot have multiple `pending states`. `CONTINUE_PENDING` is only a pending state if the final state is `RUNNING`. Otherwise, it is treated as any other state if the final state is `STOPPED`, for example.

Reply:
It might not be a bad idea to have something like
```
wait_for_pending_transition(service, pending_state, final_state)
wait_for_non_pending_transition(service, valid_initial_states, possible_final_states)
```

Reply:
That would also let us distinguish a `pending` transition from a regular transition, which from the discussion above seem like they should be treated as two separate cases.

Reply:
Okay, I understand what you are saying.

I'll update this with functions to 1. wait to see a specific state and 2. wait on a pending state